// streamr-client-protocol/dist/src/protocol/message_layer/GroupKeyRequest.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const validations_1 = require("../../utils/validations");
const GroupKeyMessage_1 = __importDefault(require("./GroupKeyMessage"));
const StreamMessage_1 = __importDefault(require("./StreamMessage"));
class GroupKeyRequest extends GroupKeyMessage_1.default {
    constructor({ requestId, streamId, rsaPublicKey, groupKeyIds }) {
        super(streamId, StreamMessage_1.default.MESSAGE_TYPES.GROUP_KEY_REQUEST);
        (0, validations_1.validateIsString)('requestId', requestId);
        this.requestId = requestId;
        (0, validations_1.validateIsString)('rsaPublicKey', rsaPublicKey);
        this.rsaPublicKey = rsaPublicKey;
        (0, validations_1.validateIsArray)('groupKeyIds', groupKeyIds);
        this.groupKeyIds = groupKeyIds;
    }
    toArray() {
        return [this.requestId, this.streamId, this.rsaPublicKey, this.groupKeyIds];
    }
    static fromArray(args) {
        const [requestId, streamId, rsaPublicKey, groupKeyIds] = args;
        return new GroupKeyRequest({
            requestId,
            streamId,
            rsaPublicKey,
            groupKeyIds,
        });
    }
    static is(streamMessage) {
        return streamMessage.messageType === StreamMessage_1.default.MESSAGE_TYPES.GROUP_KEY_REQUEST;
    }
}
exports.default = GroupKeyRequest;
GroupKeyMessage_1.default.classByMessageType[StreamMessage_1.default.MESSAGE_TYPES.GROUP_KEY_REQUEST] = GroupKeyRequest;
