// streamr-client-protocol/dist/src/protocol/message_layer/StreamMessageSerializerV31.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const StreamMessage_1 = __importDefault(require("./StreamMessage"));
const MessageRef_1 = __importDefault(require("./MessageRef"));
const MessageIDStrict_1 = __importDefault(require("./MessageIDStrict"));
const Serializer_1 = require("../../Serializer");
const VERSION = 31;
class StreamMessageSerializerV31 extends Serializer_1.Serializer {
    toArray(streamMessage) {
        return [
            VERSION,
            streamMessage.messageId.toArray(),
            streamMessage.prevMsgRef ? streamMessage.prevMsgRef.toArray() : null,
            streamMessage.messageType,
            streamMessage.encryptionType,
            streamMessage.serializedContent,
            streamMessage.signatureType,
            streamMessage.signature,
        ];
    }
    fromArray(arr) {
        const [version, // eslint-disable-line @typescript-eslint/no-unused-vars
        messageIdArr, prevMsgRefArr, messageType, encryptionType, serializedContent, signatureType, signature,] = arr;
        return new StreamMessage_1.default({
            messageId: MessageIDStrict_1.default.fromArray(messageIdArr),
            prevMsgRef: prevMsgRefArr ? MessageRef_1.default.fromArray(prevMsgRefArr) : null,
            content: serializedContent,
            messageType,
            contentType: StreamMessage_1.default.CONTENT_TYPES.JSON,
            encryptionType,
            signatureType,
            signature,
        });
    }
}
exports.default = StreamMessageSerializerV31;
StreamMessage_1.default.registerSerializer(VERSION, new StreamMessageSerializerV31());
