// streamr-client-protocol/dist/src/protocol/control_layer/resend_request/ResendFromRequest.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ControlMessage_1 = __importDefault(require("../ControlMessage"));
const validations_1 = require("../../../utils/validations");
const MessageRef_1 = __importDefault(require("../../message_layer/MessageRef"));
class ResendFromRequest extends ControlMessage_1.default {
    constructor({ version = ControlMessage_1.default.LATEST_VERSION, requestId, streamId, streamPartition, fromMsgRef, publisherId, sessionToken }) {
        super(version, ControlMessage_1.default.TYPES.ResendFromRequest, requestId);
        (0, validations_1.validateIsNotEmptyString)('streamId', streamId);
        (0, validations_1.validateIsNotNegativeInteger)('streamPartition', streamPartition);
        (0, validations_1.validateIsString)('publisherId', publisherId, true);
        (0, validations_1.validateIsString)('sessionToken', sessionToken, true);
        (0, validations_1.validateIsType)('fromMsgRef', fromMsgRef, 'MessageRef', MessageRef_1.default);
        this.streamId = streamId;
        this.streamPartition = streamPartition;
        this.fromMsgRef = fromMsgRef;
        this.publisherId = publisherId;
        this.sessionToken = sessionToken;
        (0, validations_1.validateIsNotEmptyString)('requestId', requestId); // unnecessary line once V1 is dropped
    }
}
exports.default = ResendFromRequest;
